pkg/cmd/codespace/ssh.go

```diff
@@ -147,6 +171,158 @@ func (a *App) SSH(ctx context.Context, sshArgs []string, opts sshOptions) (err e
 	}
 }
 
+func (a *App) printOpenSSHConfig(ctx context.Context, opts configOptions) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	var err error
+	var codespaces []*api.Codespace
+	if opts.codespace == "" {
+		a.StartProgressIndicatorWithLabel("Fetching codespaces")
+		codespaces, err = a.apiClient.ListCodespaces(ctx, -1)
+		a.StopProgressIndicator()
+	} else {
+		var codespace *api.Codespace
+		codespace, err = getOrChooseCodespace(ctx, a.apiClient, opts.codespace)
+		codespaces = []*api.Codespace{codespace}
+	}
+	if err != nil {
+		return fmt.Errorf("error getting codespace info: %w", err)
+	}
+
+	sshUsers := make(chan sshResult)
+	fetches := 0
+	var status error
+	for _, cs := range codespaces {
+		if cs.State != "Available" && opts.codespace == "" {
+			fmt.Fprintf(os.Stderr, "skipping unavailable codespace %s: %s\n", cs.Name, cs.State)
+			status = cmdutil.SilentError
+			continue
+		}
+
+		cs := cs
+		fetches += 1
+		go func() {
+			result := sshResult{}
```

Comment:
The counter and deferred select is a very atypical way to express a parallel loop in Go. Prefer this form:

```
ch = make chan(...)

// start all the tasks
var wg sync.WaitGroup // (this is a counting semaphore of active tasks)
for {
    wg.Add(1)
    go func() {
       defer wg.Done()
       ch <- ...result/error of subtask...
    }()
}

// wait for all the tasks to finish.
go func() {
    wg.Wait()
    close(ch) // (this terminates the range loop below)
}()

// process the results of all tasks.
for x := range ch {
   ...process each result/error x...
}
```

It is curiously laborious (some call 'go' the "goto of concurrency"), but this pattern is instantly familiar to Go programmers. No doubt a million competing library implementations will appear once the generics floodgates open. :)

As @mislav explained, there's no need to think about how to abort wg.Wait due to context cancellation if all the subgoroutines finish quickly in this case.

Reply:
OK. Thanks for the guidance! I did [this](https://github.com/cli/cli/pull/4915/commits/38ae7bb903423a3012454b4cb0e931a52a4ba39a)

I have a question though: Previously each goroutine did the deferred select, which means that if the channel were full (because the main coroutine stopped reading from it on cancellation), they were safe from blocking on the `sshUsers <- result` - they'd see the `ctx.Done()` and exit.

Was that too paranoid? Now, the `wg.Done()` doesn't happen until the defer runs, which is _after_ `sshUsers <- result`, which could theoretically block. Do I have to reserve enough space in the channel to ensure this will never happen? Or is it more that if cancellation does happen, each goroutine will notice, return an error on the channel, and the main `for result := range sshUsers` loop will just run and exit normally?

Reply:
> Now, the wg.Done() doesn't happen until the defer runs, which is after sshUsers <- result, which could theoretically block. Do I have to reserve enough space in the channel to ensure this will never happen? Or is it more that if cancellation does happen, each goroutine will notice, return an error on the channel, and the main for result := range sshUsers loop will just run and exit normally?

Well spotted! The contract around the unbuffered channel is that the main goroutine must drain it on all execution paths, otherwise, as you say, all unfinished subtasks will stay blocked in send forever. You could do this by ensuring that any break/return in the receiver loop is preceded by a drain statement (`for range ch {}`), but that's a messy invariant. You could make a simpler receiver loop that appends channel items to a slice, but this means you won't start processing the first item until the last has arrived. The best solution, when you know the number of elements beforehand is, as you say, to allocate a channel buffer large enough to hold them all.

Reply:
Thanks! Should be fixed now.